proposals/init.md

```diff
@@ -0,0 +1,581 @@
+exactly what it seems. It will not protect against the following circumstances:
+
+1. Reflection over `public` members
+1. The use of `dynamic` 
+1. Compilers that don't recognize modreqs
+
+It should also be considered that, when we complete the IL verification rules 
+for .NET 5, `init` will be one of those rules. That means extra enforcement 
+will be gained from simply verifying compilers emitting verifiable IL.
+
+The primary languages for .NET (C#, F# and VB) will all be updated to 
+recognize these `init` accessors. Hence the only realistic scenario here is 
+when a C# 9 compiler emits `init` properties and they are seen by an older 
+toolset such as C# 8, VB 15, etc ... C# 8. That is the trade off to consider
+and weigh against binary compatibility.
+
+**Note**
+This discussion primarily applies to members only, not to fields. While `init`
+fields were rejected by LDM they are still interesting to consider for the 
+modreq vs. attribute discussion. The `init` feature for fields is a relaxation
+of the existing restriction of `readonly`. That means if we emit the fields as
+`readonly` + an attribute there is no risk of older compilers mis-using the 
+field because they would already recognize `readonly`. Hence using a modreq here
+doesn't add any extra protection.
+
+**Resolution**
+The feature will use a modreq to encode the property `init` setter. The
+compelling factors were (in no particular order):
+
+* Desire to discourage older compilers from violating `init` semantics
+* Desire to make adding or removing `init` in a `virtual` declaration or 
+`interface` both a source and binary breaking change.
+
+Given there was also no significant support for removing `init` to be a 
+binary compatible change it made the choice of using modreq straight forward.
+
+### init vs. initonly
+There were three syntax forms which got significant consideration during our
+LDM meeting:
+
+```cs
+// 1. Use init 
+int Option1 { get; init; }
+// 2. Use init set
+int Option2 { get; init set; }
+// 3. Use initonly
+int Option3 { get; initonly; }
+```
+
+**Resolution**
+There was no syntax which was overwhelmingly favored in LDM.
+
+One point which got significant attention was how the choice of syntax would
+impact our ability to do `init` members as a general feature in the future.
+Choosing option 1 would mean that it would be difficult to define a property
+which had a `init` style `get` method in the future. Eventually it was decided
+that if we decided to go forward with general `init` members in future we could
+allow `init` to be a modifier in the property accessor list as well as a short
+hand for `init set`. Essentially the following two declarations would be
+identical.
+
+```cs
+int Property1 { get; init; }
+int Property1 { get; init set; }
+```
+
+The decision was made to move forward with `init` as a standalone accessor in
+the property accessor list.
+
+### Warn on failed init
```

Comment:
Any plans to add some option or another keyword that would force property to be set in initializer or constructor?

Reply:
We've touched on that topic in recent discussions, but haven't made any plans for that yet.